rasa/utils/tensorflow/layers.py

```diff
@@ -815,19 +852,52 @@ def _loss_softmax(
         sim_neg_li: tf.Tensor,
         mask: Optional[tf.Tensor],
     ) -> tf.Tensor:
-        """Define softmax loss."""
-
-        logits = tf.concat(
-            [sim_pos, sim_neg_il, sim_neg_ll, sim_neg_ii, sim_neg_li], axis=-1
-        )
+        """Define cross entropy loss."""
+        # Similarity terms between input and label should be optimized relative
+        # to each other and hence use them as logits for softmax term
+        softmax_logits = tf.concat([sim_pos, sim_neg_il, sim_neg_li], axis=-1)
+
+        if not self.constrain_similarities:
+            # Concatenate other similarity terms as well. Due to this,
+            # similarity values between input and label may not be
+            # approximately bounded in a defined range.
+            softmax_logits = tf.concat(
+                [softmax_logits, sim_neg_ii, sim_neg_ll], axis=-1
+            )
 
         # create label_ids for softmax
-        label_ids = tf.zeros_like(logits[..., 0], tf.int32)
+        softmax_label_ids = tf.zeros_like(softmax_logits[..., 0], tf.int32)
 
-        loss = tf.nn.sparse_softmax_cross_entropy_with_logits(
-            labels=label_ids, logits=logits
+        softmax_loss = tf.nn.sparse_softmax_cross_entropy_with_logits(
+            labels=softmax_label_ids, logits=softmax_logits
         )
 
+        loss = softmax_loss
+
+        if self.constrain_similarities:
+            # Constrain similarity values in a range by applying sigmoid
+            # on them individually so that they saturate at extreme values.
+            sigmoid_logits = tf.concat(
+                [sim_pos, sim_neg_il, sim_neg_ll, sim_neg_ii, sim_neg_li], axis=-1
+            )
+
+            sigmoid_labels = tf.concat(
+                [
+                    tf.expand_dims(
+                        tf.ones_like(sigmoid_logits[..., 0], tf.float32), -1
+                    ),
+                    tf.zeros_like(sigmoid_logits[..., 1:], tf.float32),
+                ],
```

Comment:
this can be simplified, also you don't need explicit `dtype` since `tf._like` will pick a type automatically the same as for inputs
```suggestion
                [tf.ones_like(sigmoid_logits[..., :1]), tf.zeros_like(sigmoid_logits[..., 1:])],
```

Reply:
Ofcourse, don't know why I complicated this op 🙈